lib/libc/gen/readpassphrase.3
.\"	$OpenBSD: readpassphrase.3,v 1.17 2007/05/31 19:19:28 jmc Exp $
.\"
.\" Copyright (c) 2000, 2002 Todd C. Miller <Todd.Miller@courtesan.com>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Sponsored in part by the Defense Advanced Research Projects
.\" Agency (DARPA) and Air Force Research Laboratory, Air Force
.\" Materiel Command, USAF, under agreement number F39502-99-1-0512.
.\"
.\" $FreeBSD$
.\"
.Dd May 31, 2007
.Dt READPASSPHRASE 3
.Os
.Sh NAME
.Nm readpassphrase
.Nd get a passphrase from the user
.Sh SYNOPSIS
.In readpassphrase.h
.Ft "char *"
.Fn readpassphrase "const char *prompt" "char *buf" "size_t bufsiz" "int flags"
.Ft "char *"
.Fn readpassphraseat "int ttyfd" "const char *prompt" "char *buf" "size_t bufsiz" "int flags"
.Sh DESCRIPTION
The
.Fn readpassphrase
function displays a prompt to, and reads in a passphrase from,
.Pa /dev/tty .
If this file is inaccessible
and the
.Dv RPP_REQUIRE_TTY
flag is not set,
.Fn readpassphrase
displays the prompt on the standard error output and reads from the standard
input.
In this case it is generally not possible to turn off echo.
.Pp
Up to
.Fa bufsiz
\- 1 characters (one is for the
.Dv NUL )
are read into the provided buffer
.Fa buf .
Any additional
characters and the terminating newline (or return) character are discarded.
.Pp
The
.Fn readpassphraseat
function is equivalent to
.Fn readpassphrase ,
but allows the user to specify a
.Fa ttyfd
terminal type device for prompting and password retrieval.
Programs in
.Xr capsicum 4
capability mode can pass a pre-opened
.Pa /dev/tty
file descriptor into
.Fn readpassphraseat
to avoid triggering a capability violation.
.Pp
The
.Fn readpassphrase
and
.Fn readpassphraseat
functions
take the following optional
.Fa flags :
.Pp
.Bl -tag -width ".Dv RPP_REQUIRE_TTY" -compact
.It Dv RPP_ECHO_OFF
turn off echo (default behavior)
.It Dv RPP_ECHO_ON
leave echo on
.It Dv RPP_REQUIRE_TTY
fail if there is no tty
.It Dv RPP_FORCELOWER
force input to lower case
.It Dv RPP_FORCEUPPER
force input to upper case
.It Dv RPP_SEVENBIT
strip the high bit from input
.It Dv RPP_STDIN
force read of passphrase from stdin
.El
.Pp
If
.Dv AT_STDIN
is assigned to
.Fa ttyfd ,
the passphrase will be force read from stdin.
This is equivalent to specifying the
.Dv RPP_STDIN
flag.
.Pp
The calling process should zero the passphrase as soon as possible to
avoid leaving the cleartext passphrase visible in the process's address
space.
.Sh RETURN VALUES
Upon successful completion,
.Fn readpassphrase
and
.Fn readpassphraseat
return a pointer to the NUL-terminated passphrase.
If an error is encountered, the terminal state is restored and
a
.Dv NULL
pointer is returned.
.Sh FILES
.Bl -tag -width ".Pa /dev/tty" -compact
.It Pa /dev/tty
.El
.Sh EXAMPLES
The following code fragment will read a passphrase from
.Pa /dev/tty
into the buffer
.Fa passbuf .
.Bd -literal -offset indent
char passbuf[1024];

\&...

if (readpassphrase("Response: ", passbuf, sizeof(passbuf),
    RPP_REQUIRE_TTY) == NULL)
	errx(1, "unable to read passphrase");

if (compare(transform(passbuf), epass) != 0)
	errx(1, "bad passphrase");

\&...

memset(passbuf, 0, sizeof(passbuf));
.Ed
.Sh ERRORS
.Bl -tag -width Er
.It Bq Er EINTR
The passphrase reading was interrupted by a signal.
.It Bq Er EINVAL
The
.Ar bufsiz
argument was zero.
.It Bq Er EIO
The process is a member of a background process attempting to read
from its controlling terminal, the process is ignoring or blocking
the
.Dv SIGTTIN
signal, or the process group is orphaned.
.It Bq Er EMFILE
The process has already reached its limit for open file descriptors.
.It Bq Er ENFILE
The system file table is full.
.It Bq Er ENOTTY
There is no controlling terminal and the
.Dv RPP_REQUIRE_TTY
flag was specified.
This error is only applicable to
.Fn readpassphrase .
.It Bq Er ENOTTY
The provided
.Fa ttyfd
is an invalid controlling terminal.
This error is only applicable to
.Fn readpassphraseat .
.El
.Sh SIGNALS
The
.Fn readpassphrase
function
will catch the following signals:
.Bd -literal -offset indent
SIGALRM		SIGHUP		SIGINT
SIGPIPE		SIGQUIT		SIGTERM
SIGTSTP		SIGTTIN		SIGTTOU
.Ed
.Pp
When one of the above signals is intercepted, terminal echo will
be restored if it had previously been turned off.
If a signal handler was installed for the signal when
.Fn readpassphrase
was called, that handler is then executed.
If no handler was previously installed for the signal then the
default action is taken as per
.Xr sigaction 2 .
.Pp
The
.Dv SIGTSTP , SIGTTIN
and
.Dv SIGTTOU
signals (stop signals generated from keyboard or due to terminal I/O
from a background process) are treated specially.
When the process is resumed after it has been stopped,
.Fn readpassphrase
will reprint the prompt and the user may then enter a passphrase.
.Sh SEE ALSO
.Xr sigaction 2 ,
.Xr getpass 3 ,
.Xr capsicum 4
.Sh STANDARDS
The
.Fn readpassphrase
function is an
extension and should not be used if portability is desired.
.Sh HISTORY
The
.Fn readpassphrase
function first appeared in
.Fx 4.6
and
.Ox 2.9 .


lib/libc/gen/readpassphrase.c
/*	$OpenBSD: readpassphrase.c,v 1.24 2013/11/24 23:51:29 deraadt Exp $	*/

/*
 * Copyright (c) 2000-2002, 2007, 2010
 *	Todd C. Miller <Todd.Miller@courtesan.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Sponsored in part by the Defense Advanced Research Projects
 * Agency (DARPA) and Air Force Research Laboratory, Air Force
 * Materiel Command, USAF, under agreement number F39502-99-1-0512.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include "namespace.h"
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <paths.h>
#include <pwd.h>
#include <signal.h>
#include <string.h>
#include <termios.h>
#include <unistd.h>
#include <readpassphrase.h>
#include "un-namespace.h"
#include "libc_private.h"

static volatile sig_atomic_t signo[NSIG];

static void handler(int);

char *
readpassphraseat(int ttyfd, const char *prompt, char *buf, size_t bufsiz, int flags)
{
	ssize_t nr;
	int input, output, save_errno, i, need_restart, input_is_tty;
	char ch, *p, *end;
	struct termios term, oterm;
	struct sigaction sa, savealrm, saveint, savehup, savequit, saveterm;
	struct sigaction savetstp, savettin, savettou, savepipe;

	/* I suppose we could alloc on demand in this case (XXX). */
	if (bufsiz == 0) {
		errno = EINVAL;
		return(NULL);
	}

	/* Passing AT_STDIN in ttyfd is identical to RPP_STDIN. */
	if (ttyfd == AT_STDIN ||
	    ttyfd == STDIN_FILENO ||
	    ttyfd == STDOUT_FILENO ||
	    ttyfd == STDERR_FILENO)
		flags |= RPP_STDIN;

restart:
	for (i = 0; i < NSIG; i++)
		signo[i] = 0;
	nr = -1;
	save_errno = 0;
	need_restart = 0;
	/*
	 * Read and write to terminal type device if valid.  If not, read
	 * from stdin and write to stderr.
	 */
	input_is_tty = 0;
	if ((flags & RPP_STDIN) == 0) {
		if (isatty(ttyfd) != 1) {
			errno = ENOTTY;
			return (NULL);
		}
		input_is_tty = 1;
		input = output = ttyfd;
	} else {
		input = STDIN_FILENO;
		output = STDERR_FILENO;
	}

	/*
	 * Turn off echo if possible.
	 * If we are using a tty but are not the foreground pgrp this will
	 * generate SIGTTOU, so do it *before* installing the signal handlers.
	 */
	if (input_is_tty && tcgetattr(input, &oterm) == 0) {
		memcpy(&term, &oterm, sizeof(term));
		if (!(flags & RPP_ECHO_ON))
			term.c_lflag &= ~(ECHO | ECHONL);
		if (term.c_cc[VSTATUS] != _POSIX_VDISABLE)
			term.c_cc[VSTATUS] = _POSIX_VDISABLE;
		(void)tcsetattr(input, TCSAFLUSH|TCSASOFT, &term);
	} else {
		memset(&term, 0, sizeof(term));
		term.c_lflag |= ECHO;
		memset(&oterm, 0, sizeof(oterm));
		oterm.c_lflag |= ECHO;
	}

	/*
	 * Catch signals that would otherwise cause the user to end
	 * up with echo turned off in the shell.  Don't worry about
	 * things like SIGXCPU and SIGVTALRM for now.
	 */
	sigemptyset(&sa.sa_mask);
	sa.sa_flags = 0;		/* don't restart system calls */
	sa.sa_handler = handler;
	(void)__libc_sigaction(SIGALRM, &sa, &savealrm);
	(void)__libc_sigaction(SIGHUP, &sa, &savehup);
	(void)__libc_sigaction(SIGINT, &sa, &saveint);
	(void)__libc_sigaction(SIGPIPE, &sa, &savepipe);
	(void)__libc_sigaction(SIGQUIT, &sa, &savequit);
	(void)__libc_sigaction(SIGTERM, &sa, &saveterm);
	(void)__libc_sigaction(SIGTSTP, &sa, &savetstp);
	(void)__libc_sigaction(SIGTTIN, &sa, &savettin);
	(void)__libc_sigaction(SIGTTOU, &sa, &savettou);

	if (!(flags & RPP_STDIN))
		(void)_write(output, prompt, strlen(prompt));
	end = buf + bufsiz - 1;
	p = buf;
	while ((nr = _read(input, &ch, 1)) == 1 && ch != '\n' && ch != '\r') {
		if (p < end) {
			if ((flags & RPP_SEVENBIT))
				ch &= 0x7f;
			if (isalpha((unsigned char)ch)) {
				if ((flags & RPP_FORCELOWER))
					ch = (char)tolower((unsigned char)ch);
				if ((flags & RPP_FORCEUPPER))
					ch = (char)toupper((unsigned char)ch);
			}
			*p++ = ch;
		}
	}
	*p = '\0';
	save_errno = errno;
	if (!(term.c_lflag & ECHO))
		(void)_write(output, "\n", 1);

	/* Restore old terminal settings and signals. */
	if (memcmp(&term, &oterm, sizeof(term)) != 0) {
		while (tcsetattr(input, TCSAFLUSH|TCSASOFT, &oterm) == -1 &&
		    errno == EINTR && !signo[SIGTTOU])
			continue;
	}
	(void)__libc_sigaction(SIGALRM, &savealrm, NULL);
	(void)__libc_sigaction(SIGHUP, &savehup, NULL);
	(void)__libc_sigaction(SIGINT, &saveint, NULL);
	(void)__libc_sigaction(SIGQUIT, &savequit, NULL);
	(void)__libc_sigaction(SIGPIPE, &savepipe, NULL);
	(void)__libc_sigaction(SIGTERM, &saveterm, NULL);
	(void)__libc_sigaction(SIGTSTP, &savetstp, NULL);
	(void)__libc_sigaction(SIGTTIN, &savettin, NULL);
	(void)__libc_sigaction(SIGTTOU, &savettou, NULL);

	/*
	 * If we were interrupted by a signal, resend it to ourselves
	 * now that we have restored the signal handlers.
	 */
	for (i = 0; i < NSIG; i++) {
		if (signo[i]) {
			kill(getpid(), i);
			switch (i) {
			case SIGTSTP:
			case SIGTTIN:
			case SIGTTOU:
				need_restart = 1;
			}
		}
	}
	if (need_restart)
		goto restart;

	if (save_errno)
		errno = save_errno;
	return(nr == -1 ? NULL : buf);
}

char *
readpassphrase(const char *prompt, char *buf, size_t bufsiz, int flags)
{
	int save_errno, ttyfd;
	char *error;

	ttyfd = AT_STDIN;
	/*
	 * Pass /dev/tty fd into readpassphraseat() if available.
	 * Otherwise, resort to stdin unless a tty is required.
	 */
	if ((flags & RPP_STDIN) == 0) {
		ttyfd = _open(_PATH_TTY, O_RDWR | O_CLOEXEC);
		if (ttyfd == -1) {
			if ((flags & RPP_REQUIRE_TTY) != 0) {
				errno = ENOTTY;
				return (NULL);
			}
			ttyfd = AT_STDIN;
		}
	}
	error = readpassphraseat(ttyfd, prompt, buf, bufsiz, flags);

	save_errno = errno;
	if (ttyfd != AT_STDIN)
		(void)_close(ttyfd);
	errno = save_errno;

	return (error);
}

char *
getpass(const char *prompt)
{
	static char buf[_PASSWORD_LEN + 1];

	if (readpassphrase(prompt, buf, sizeof(buf), RPP_ECHO_OFF) == NULL)
		buf[0] = '\0';
	return(buf);
}

static void handler(int s)
{

	signo[s] = 1;
}
